File: com/heapsort/theAlgorithms/theAlgorithms.java
package com.heapsort.theAlgorithms;
public class theAlgorithms{

    public int[] sort(int[] unsorted) {
        int n = unsorted.length;
        heapify(unsorted, n);
        while (n > 1) {
            swap(unsorted, 1, n--);
            siftDown(unsorted, 1, n);
        }
        return unsorted;
    }

    private static void heapify(int[] unsorted, int n) {
        for (int k = n / 2; k >= 1; k--) {
            siftDown(unsorted, k, n);
        }
    }

    private static void siftDown(int[] unsorted, int k, int n) {
        while (2 * k <= n) {
            int j = 2 * k;
            if (j < n && less(unsorted, j, j + 1)) {
                j++;
            }
            if (!less(unsorted, k, j)) {
                break;
            }
            swap(unsorted, k, j);
            k = j;
        }
    }

    private static void swap(int[] array, int idx, int idy) {
        int swap = array[idx - 1];
        array[idx - 1] = array[idy - 1];
        array[idy - 1] = swap;
    }

    private static boolean less(int[] array, int idx, int idy) {
        return array[idx - 1] < array[idy - 1];
    }
}

File: Python/heapsort/theAlgorithms/theAlgorithms.py
import random

def heapify(unsorted, index, heap_size):
    largest = index
    left_index = 2 * index + 1
    right_index = 2 * index + 2
    if left_index < heap_size and unsorted[left_index] > unsorted[largest]:
        largest = left_index

    if right_index < heap_size and unsorted[right_index] > unsorted[largest]:
        largest = right_index

    if largest != index:
        unsorted[largest], unsorted[index] = unsorted[index], unsorted[largest]
        heapify(unsorted, largest, heap_size)


def heap_sort(unsorted):
    """
    Pure implementation of the heap sort algorithm in Python
    :param collection: some mutable ordered collection with heterogeneous
    comparable items inside
    :return: the same collection ordered by ascending

    Examples:
    >>> heap_sort([0, 5, 3, 2, 2])
    [0, 2, 2, 3, 5]

    >>> heap_sort([])
    []

    >>> heap_sort([-2, -5, -45])
    [-45, -5, -2]
    """
    n = len(unsorted)
    for i in range(n // 2 - 1, -1, -1):
        heapify(unsorted, i, n)
    for i in range(n - 1, 0, -1):
        unsorted[0], unsorted[i] = unsorted[i], unsorted[0]
        heapify(unsorted, 0, i)
    return unsorted


if __name__ == "__main__":
    # Generate a list of 1000 random integers between 1 and 1000
    unsorted = [random.randint(1, 1000) for _ in range(1000)]

    # Sort the list using heap_sort
    heap_sort(unsorted)

    # The list is now sorted, no print statements included
